### PR TITLE
Fix wrong socket type for extra addresses in QUIC listeners

### PR DESCRIPTION
**Pull request by leosarra**
**Please provide a description of this PR:**
This will address a DS-exclusive issue: #48336

As I mentioend in the issue, when both DS and QUIC listeners are enabled we end up using different socket type between the address and additional_addresses.
Because we end up with both an UDP and a TCP address in the same listener, Envoy will correctly rejects the configuration.

I have updated the code to ensure a proper listener configuration.

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/networking/util/util.go [hunk 1]
```diff
@@ -171,9 +171,20 @@ func AddrStrToCidrRange(addr string) (*core.CidrRange, error) {
...
 
+// OptionalBuildOptions is a struct to hold optional parameters for BuildAddress and BuildAdditionalAddresses.
+type OptionalBuildAddressOptions struct {
+	Transport istionetworking.TransportProtocol
+}
+
 // BuildAddress returns a SocketAddress with the given ip and port or uds.
-func BuildAddress(bind string, port uint32) *core.Address {
-	address := BuildNetworkAddress(bind, port, istionetworking.TransportProtocolTCP)
+func BuildAddress(bind string, port uint32, options ...OptionalBuildAddressOptions) *core.Address {
```
Comment (leosarra) [inline pilot/pkg/networking/util/util.go:180]: I used a variadic variable to add extra build options for simplicity.
If you prefer I can add the transport socket_type as a normal parameter but I will have to update every existing use of BuildAddress and BuildAdditionalAddress to use TCP (except for that one time where QUIC listeners are needed).

**Comment by leosarra** [comment]
@kfaseela @zhlsunshine

**Comment by kfaseela** [comment]
@leosarra could you please unskip the quic tests for dual-stack and show that it passes now with the fix? https://github.com/istio/istio/blob/master/tests/integration/security/sds_ingress/quic/ingress_test.go#L111

File: pilot/pkg/networking/util/util.go [hunk 2]
```diff
@@ -171,9 +171,20 @@ func AddrStrToCidrRange(addr string) (*core.CidrRange, error) {
 	}, nil
 }
 
+// OptionalBuildOptions is a struct to hold optional parameters for BuildAddress and BuildAdditionalAddresses.
+type OptionalBuildAddressOptions struct {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/util/util.go:175]: cannot you pass the transport directly, and i also do not see the need to use `...OptionalBuildAddressOptions`

**Review by hzxuzhonghu — COMMENTED** [review]
please add a test

**Comment by leosarra** [comment]
> please add a test

I have removed the skip condition from the ingress QUIC tests that were failing on master without this fix

File: pilot/pkg/networking/util/util.go (hunk 2 above)
Comment (leosarra) [reply pilot/pkg/networking/util/util.go:175]: Are you proposing to update the tinker with the value after the address has been created?
E.g: 
```
for every extraAddress:
     extraAddress.Address.GetSocketAddress().Protocol = transport.ToEnvoyProtocol()
```

File: tests/integration/security/sds_ingress/quic/ingress_test.go [hunk 3]
```diff
@@ -79,12 +79,6 @@ func TestTlsGatewaysWithQUIC(t *testing.T) {
 		RequiresSingleCluster().
 		Features("security.ingress.quic.sds.tls").
 		Run(func(t framework.TestContext) {
-			if t.Settings().EnableDualStack {
```
Comment (kfaseela) [inline tests/integration/security/sds_ingress/quic/ingress_test.go:82]: Line 107 also needs to be unskipped

Comment (leosarra) [reply tests/integration/security/sds_ingress/quic/ingress_test.go:82]: Thank you. Done

**Comment by kfaseela** [comment]
> > @leosarra could you please unskip the quic tests for dual-stack and show that it passes now with the fix? https://github.com/istio/istio/blob/master/tests/integration/security/sds_ingress/quic/ingress_test.go#L111
> 
> > please add a test
> 
> I have removed the skip condition from the ingress QUIC tests that were failing on master without this fix

The presubmit tests won't usually run this test. If you want to make sure if runs fine before merging the PR, you can temporarily edit this line and see it passes :) https://github.com/istio/istio/blob/master/tests/integration/tests.mk#L129

**Comment by leosarra** [comment]
> The presubmit tests won't usually run this test. If you want to make sure if runs fine before merging the PR, you can temporarily edit this line and see it passes :) https://github.com/istio/istio/blob/master/tests/integration/tests.mk#L129

Ah, I see :)
This time I have tested the QUIC tls/mtls tests locally by using a personal build.

**Comment by kfaseela** [comment]
LGTM from test perspective. Would leave to @hzxuzhonghu @zhlsunshine and @howardjohn for final approval

**Review by howardjohn — APPROVED** [review]

**Pull request merged**